### PR TITLE
Add `ElectraForCausalLM` -> Enable Electra encoder-decoder model

### LINKED ISSUE #14269: Add `ElectraForCausalLM` to enable constructing Electra-based `EncoderDecoderModel`
# 🚀 Feature request

Add `ElectraForCausalLM` model to enable constructing `EncoderDecoderModel` based on Electra.

## Motivation

Currently, there are many encoder-only models, such as BERT or RoBERTA, which have already been supported for encoder-decoder settings. Even though Electra is pre-trained in a different fashion, I found out the fine-tuning an Electra2Electra model works as well.
Also, there are publicly available checkpoints for a small configuration of the model and thus one can try to train a model which is an order of magnitude smaller than e.g. Bert2Bert or RoBERTa2RoBERTa.

## Your contribution

I have already prepared an implementation for myself. If there's a desire to add this module I'll be happy to work on the PR. :]

### PR DESCRIPTION
**Pull request by stancld**
# What does this PR do?

This PR adds `ElectraForCausalLM` and thus enable to use `Electra` model in the encoder-decoder setting.

Fixes #14269

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten 

**Additional context:** I've already tried to fine-tune Electra model as an encoder-decoder model on our use case and it proved to be working.

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
This looks good to me - we'll need to update the documentation to the new format before merging, and pinging @patrickvonplaten for an additional review.

Thanks a lot for your PR @stancld, very appreciated!

File: src/transformers/models/electra/modeling_electra.py [hunk 1]
```diff
@@ -1078,14 +1115,14 @@ def forward(
...
             input_ids,
-            attention_mask,
-            token_type_ids,
-            position_ids,
-            head_mask,
-            inputs_embeds,
-            output_attentions,
-            output_hidden_states,
-            return_dict,
+            attention_mask=attention_mask,
```
Comment (patrickvonplaten) [inline src/transformers/models/electra/modeling_electra.py:1118]: nice!

**Comment by patrickvonplaten** [comment]
This looks good to me as well! Regarding the docs, could you try to follow @sgugger 's suggestion [here](https://github.com/huggingface/transformers/pull/13967#issuecomment-999565664) on how to fix it? 

Let me know if you need any help :-)

File: src/transformers/models/electra/modeling_electra.py [hunk 2]
```diff
@@ -691,6 +692,15 @@ def _set_gradient_checkpointing(self, module, value=False):
         if isinstance(module, ElectraEncoder):
             module.gradient_checkpointing = value
 
+    def update_keys_to_ignore(self, config, del_keys_to_ignore):
```
Comment (patrickvonplaten) [inline src/transformers/models/electra/modeling_electra.py:695]: is this necessary? is it ever called actually?

File: src/transformers/models/electra/modeling_electra.py [hunk 3]
```diff
@@ -1481,3 +1518,151 @@ def forward(
...
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        labels=None,
+        past_key_values=None,
+        use_cache=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
+        r"""
```
Comment (sgugger) [inline src/transformers/models/electra/modeling_electra.py:1564]: This docstring needs to be converted to Markdown. You might have an easier time copying it from where you took it on the master branch than trying to use our tools to convert it however.

**Review by sgugger — APPROVED** [review]
Thanks for adding this model!

File: src/transformers/models/electra/modeling_electra.py (hunk 2 above)
Comment (stancld) [reply src/transformers/models/electra/modeling_electra.py:695]: Actually looks like it's not necessary. I'll remove it then.

**Comment by patrickvonplaten** [comment]
Merging as failing test has been fixed on master and other failures are unrelated. Thanks a lot for you work @stancld !

**Pull request merged**